File: Criacional/Abstract Factory/AbstractFactory01.js
//PRODUTOS ABSTRADOS
class GatewayPagamento{
    autorizar(valor){
        throw new Error("autorizar() deve ser implementado.")
    }
}

class Recibo{
    gerar(valor){
        throw new Error("gerar() deve ser implementado.")
    }
}

//PRODUTOS CONCRETOS
class GatewayPayPal extends GatewayPagamento{
    autorizar(valor){
        return `Paypal: Pagamento de R${valor.toFixed(2)} autorizado.` 
    }
}

class ReciboPayPal extends Recibo{
    gerar(valor){
        return `PayPal: Recibo do pagamento de R${valor.toFixed(2)}.`
    }

}


class GatewayMercadoPago extends GatewayPagamento{
    autorizar(valor){
        return `Mercado Pago: Pagamento de R${valor.toFixed(2)} autorizado.`
    }
}

class ReciboMercadoPago extends Recibo{
    gerar(valor){
        return `Mercado Pago: Recibo da pagamento de R${valor.toFixed(2)}.`
    }
}


//  FÁBRICA ABSTRATA
class FabricaPagamento{
    criatGateway(){
        throw new Error("criarGateway() deve ser implementado")
    }

    criarRecibo(){
        throw new Error("criarRecibo() deve ser implementado")
    }
}


//FABRICA CONCRETA

class FabricaPayPal{
    criarGateway(){
        return new GatewayPayPal;
    }

    criarRecibo(){
        return new ReciboPayPal;
    }
}


class FabricaMercadoPago{
    criarGateway(){
        return new GatewayMercadoPago;
    }

    criarRecibo(){
        return new ReciboMercadoPago
    }
}



//MAIN
function finalizarCompra(fabrica, valor){
    const gateway = fabrica.criarGateway();
    const recibo = fabrica.criarRecibo();

    console.log(gateway.autorizar(valor));
    
    console.log(recibo.gerar(valor));
}

finalizarCompra(new FabricaPayPal(), 100.99 )
console.log("*************************")
finalizarCompra(new FabricaMercadoPago(),3100.99 )
console.log("*************************")
finalizarCompra(new FabricaPayPal(), 180.99 )
console.log("*************************")
finalizarCompra(new FabricaMercadoPago(), 1008.99 )

File: Criacional/Abstract Factory/AbstractFactory01.py
#PRODUTOS ABSTRATOS

class GatewayPagamento:
    def autorizar(self, valor):
        raise NotImplementedError("autorizar() deve ser implementado")


class Recibo:
    def gerar(self, valor):
        raise NotImplementedError("gerar() deve ser implementado")

#PRODUTOS CONCRETOS - PAYPAL

class GatewayPayPal(GatewayPagamento):
    def autorizar(self, valor):
        return f"PayPal: Pagamento de R$ {valor:.2f} autorizado"

class ReciboPayPal(Recibo):
    def gerar(self, valor):
        return f"PayPal: Recibo do pagamento de R$ {valor:.2f}."

#PRODUTOS CONCRETOS - MERCADOPAGO

class GatewayMercadoPago(GatewayPagamento):
    def autorizar(self, valor):
        return f"Mercado pago: Pagagamento de R$ {valor:.2f} autorizado"


class ReciboMercadoPago(Recibo):
    def gerar(self, valor):
        return f"Mercado Pago: Recibo do pagamento de R$ {valor:.2f}."

#FABRICA ABSTRATA
class FabricaPagamento:
    def criarGateway(self):
        raise NotImplementedError("criarGateway() deve ser implementado")

    def criarRecibo(self):
        raise NotImplementedError("criarRecibo() deve ser implementado")

    
#FABRICA CONCRETA
class FabricaPayPal(FabricaPagamento):
    def criarGateway(self):
        return GatewayPayPal()

    def criarRecibo(self):
        return ReciboPayPal()

class FabricaMercadoPago(FabricaPagamento):
    def criarGateway(self):
        return GatewayMercadoPago()

    def criarRecibo(self):
        return ReciboMercadoPago()


#MAIN

def finalizarCompra(fabrica, valor):
    gateway = fabrica.criarGateway()
    recibo = fabrica.criarRecibo()

    print(gateway.autorizar(valor))
    print(recibo.gerar(valor))


finalizarCompra(FabricaPayPal(), 100.99 )
print("*************************")
finalizarCompra(FabricaMercadoPago(),3100.99 )
print("*************************")
finalizarCompra(FabricaPayPal(), 180.99 )
print("*************************")
finalizarCompra(FabricaMercadoPago(), 1008.99 )